ts/src/profiler.ts

```diff
@@ -13,47 +13,292 @@
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    let requestError: Error|undefined = undefined;
+    let retryRequest = true;
+    try {
+      const [body, response] = await this.request(options);
+      if (isErrorResponseCode(response.statusCode)) {
+        retryRequest = isRetriableResponseCode(response.statusCode);
+        requestError =
+            new Error('Error creating profile: ' + response.statusMessage);
+        this.logger.debug(requestError);
+      } else {
+        return body;
+      }
+    } catch (err) {
+      retryRequest =
+          isRetriableError(err) || isRetriableResponseCode(err.statusCode);
+      requestError = new Error('Error creating profile: ' + err.toString());
+      this.logger.debug(requestError);
+    }
+    if (retryRequest) {
+      // TODO: check response to see if response specifies a backoff.
+      // TODO: implement exponential backoff.
+      await delay(this.config.backoffMillis);
+      return this.createProfile();
+    }
+    throw requestError;
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, like a problem collecting or uploading the
+   * profile, an error will be logged at the debug level, but otherwise ignored.
+   *
+   * Public to allow for testing.
+   *
+   * @param prof
+   */
+  async profileAndUpload(prof: RequestProfile): Promise<void> {
+    try {
+      prof = await this.profile(prof);
+    } catch (err) {
+      this.logger.debug('Error collecting profile: ' + err.toString());
+      return;
+    }
+    const options = {
+      method: 'PATCH',
+      uri: API + '/' + prof.name,
+      body: prof,
+      json: true,
+    };
+    try {
+      const [body, response] = await this.request(options);
+      if (isErrorResponseCode(response.statusCode)) {
+        this.logger.debug('Error uploading profile: ' + response.statusMessage);
+      }
+    } catch (err) {
+      this.logger.debug('Error uploading profile: ' + err.toString());
+    }
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, for example the profileType is not
+   * recognized or profiling is disabled for the specified profileType, an
+   * error will be thrown.
+   *
+   * Public to allow for testing.
+   *
+   * @param prof
+   */
+  async profile(prof: RequestProfile): Promise<RequestProfile> {
+    switch (prof.profileType) {
+      case ProfileTypes.Wall:
+        return await this.writeTimeProfile(prof);
+      case ProfileTypes.Heap:
+        return this.writeHeapProfile(prof);
+      default:
+        throw new Error('Unexpected profile type ' + prof.profileType + '.');
+    }
+  }
+
+  /**
+   * Collects a time profile, converts profile to compressed, base64 encoded
+   * string, and adds profileBytes field to prof with this string.
+   *
+   * Public to allow for testing.
```

Comment:
I don't see this function being tested, is the comment out of date? Can you check other functions? I would also prefer not to have these individual functions tested as much as possible anyway and rather have the test exercise the overall behavior talking to some kind of fake profile server so that things are close to what they are going to be in production.

Reply:
This function (and other functions with this in the comment) are tested. 
(https://github.com/nolanmar511/cloud-profiler-nodejs/blob/bb77e78eb4250b5694762af286284e5be4c51771/ts/test/test-profiler.ts#L131).

cloud-trace-nodejs and cloud-debug-nodejs have unit tests (like the ones I've written), and system tests (like what you've described) -- I know I need to add system tests, but should I also get rid of unit tests?

Reply:
Ah, OK, keep then.